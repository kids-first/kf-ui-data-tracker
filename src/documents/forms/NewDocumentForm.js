import React, {useState} from 'react';
import {Amplitude} from '@amplitude/react-amplitude';
import {
  Button,
  Divider,
  Form,
  Header,
  Icon,
  Input,
  Message,
  Segment,
} from 'semantic-ui-react';
import {Formik, Field} from 'formik';
import {EditorState, convertToRaw, ContentState} from 'draft-js';
import MarkdownEditor from '../components/FileDetail/MarkdownEditor';
import SelectElement from '../components/FileDetail/SelectElement';
import draftToMarkdown from 'draftjs-to-markdown';

import {fileTypeDetail} from '../../common/enums';
import {urls} from '../../common/urls';

const NewDocumentForm = ({
  values,
  errors,
  touched,
  isSubmitting,
  isValid,
  handleBlur,
  handleChange,
  handleSubmit,
  version,
  setFieldValue,
  setFieldTouched,
  studyFiles,
}) => {
  const [editorState, setEditorState] = useState(
    EditorState.createWithContent(ContentState.createFromText('')),
  );

  return (
    <>
      <Form.Field required>
        <Header>Document Type</Header>
        <p>
          Help the DRC understand the purpose of this document by classifying it
          as the appropriate type.
        </p>
        <Header size="small">General Types</Header>
        {Object.keys(fileTypeDetail).map(item => (
          <Form.Field key={item}>
            <Field
              component={SelectElement}
              name="file_type"
              id={item}
              label={item}
            />
          </Form.Field>
        ))}
        <Header size="small">Expedited Types</Header>
        <Message icon>
          <Icon name="info circle" />
          <Message.Content>
            <Message.Header>Does not Qualify</Message.Header>
            This file does not meet requirements for any of the expedited types.
            If you'd like to have this file processed faster, please consider
            formatting the file according to one of the{' '}
            <a
              target="_blank"
              rel="noopener noreferrer"
              href={urls.expeditedFiles}
            >
              file types
            </a>
            .
          </Message.Content>
        </Message>
      </Form.Field>

      <Divider />

      <Form.Field error={touched.file_name && !!errors.file_name} required>
        <Header>Document Name</Header>
        <Input
          data-testid="name-input"
          type="text"
          name="file_name"
          placeholder="Phenotypic Data manifest for..."
          value={values.file_name}
          onBlur={handleBlur}
          onChange={handleChange}
        />
        {errors.file_name === 'exists' && (
          <Message
            negative
            icon="warning"
            header="Document Already Exists"
            content="A document with this name already exists in the study. Consider updating the existing document instead of creating a new document"
          />
        )}
        <small>
          Please provide a descriptive title without dates or adjectives such as
          "new", "updated", "final", etc.
        </small>
      </Form.Field>

      <Divider />

      <Form.Field required>
        <Header>Description</Header>
        <Segment color={touched.file_desc && errors.file_desc ? 'red' : null}>
          <MarkdownEditor
            editorState={editorState}
            onEditorStateChange={e => {
              setFieldTouched('file_desc');
              setEditorState(e);
              const rawState = convertToRaw(e.getCurrentContent());
              const mdText = draftToMarkdown(rawState);
              setFieldValue('file_desc', mdText);
            }}
          />
        </Segment>

        <Amplitude
          eventProperties={inheritedProps => ({
            ...inheritedProps,
            scope: inheritedProps.scope
              ? [...inheritedProps.scope, 'button', 'upload button']
              : ['button', 'upload button'],
          })}
        >
          {({logEvent}) => (
            <Button
              primary
              data-testid="new-file-submit"
              type="submit"
              disabled={isSubmitting || !isValid}
              loading={isSubmitting}
              onClick={() => isValid && logEvent('click')}
            >
              Create
            </Button>
          )}
        </Amplitude>
      </Form.Field>
    </>
  );
};

const FormikWrapper = ({handleSubmit, studyFiles}, ...props) => (
  <Amplitude
    eventProperties={inheritedProps => ({
      ...inheritedProps,
      scope: inheritedProps.scope
        ? [...inheritedProps.scope, 'new document form']
        : ['new document form'],
    })}
  >
    <Formik
      initialValues={{
        file_name: '',
        file_type: 'OTH',
        file_desc: '',
        file_status: 'PEN',
      }}
      validate={vals => {
        let errors = {};
        if (!vals.file_name) errors.file_name = 'required';
        if (!vals.file_desc.trim()) errors.file_desc = 'required';
        if (
          vals.file_name &&
          studyFiles.filter(
            ({node}) =>
              node.name.toLowerCase() === vals.file_name.toLowerCase(),
          ).length > 0
        ) {
          errors.file_name = 'exists';
        }
        return errors;
      }}
      onSubmit={handleSubmit}
    >
      {formikProps => (
        <Form onSubmit={formikProps.handleSubmit}>
          <NewDocumentForm
            {...props}
            {...formikProps}
            studyFiles={studyFiles}
          />
        </Form>
      )}
    </Formik>
  </Amplitude>
);

export default FormikWrapper;
