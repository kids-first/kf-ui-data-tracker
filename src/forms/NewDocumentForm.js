import React, {useState} from 'react';
import PropTypes from 'prop-types';
import {Button, GridContainer} from 'kf-uikit';
import SelectElement from '../components/FileAnnotation/SelectElement';
import {fileTypeDetail} from '../common/fileUtils';

const plachelderText = {
  SHM: {
    title: 'Biospecimen shipping manifest from ...',
    desc:
      'Shipping manifest for biospecimens shipped from LabA to CenterB on 00/00/0000 .....',
  },
  CLN: {
    title: `20XX Phenotypic data manifest for  ...`,
    desc:
      'Phenotypic data for 00 patients and 00 containing demographics, diagnoses, and ...',
  },
  SEQ: {
    title: 'Sequencing Center manifest from [Seq. Center Name] on [date]... ',
    desc: 'Manifest received from [Seq. Center Name] accounting for ...',
  },
  OTH: {
    title: '...data sheet...',
    desc: 'Data for [subject matter] containing/accounting for ...',
  },
};

/**
 * A form for uploading a new document.
 * The form will invoke the handleSubmit handle upon submission passing it
 * fileName, fileType, and fileDescription values.
 */
const NewFileForm = ({handleSubmit, handleCancel, errors}) => {
  const [fileType, setFileType] = useState();
  const [fileName, setFileName] = useState();
  const [fileDescription, setFileDescription] = useState();
  // For the uploading stage
  const [onUploading, setUploading] = useState(false);
  console.log(onUploading);
  return (
    <form
      onSubmit={ev => {
        setUploading(true);
        ev.preventDefault();
        handleSubmit(fileName, fileType, fileDescription);
      }}
      className="Form"
    >
      <GridContainer collapsed="rows">
        <label
          className="row-1 cell-3 text-right Form--Label Form--Label-required"
          htmlFor="file_name"
        >
          Document Title:
        </label>
        <input
          data-testid="name-input"
          className="Form--Input row-1 cell-7"
          type="text"
          name="file_name"
          placeholder={fileType ? plachelderText[fileType].title : ''}
          defaultValue={fileName}
          onChange={e => setFileName(e.target.value)}
        />
        <label
          className="row-2 cell-3 text-right Form--Label Form--Label-required"
          htmlFor="file_type"
        >
          Document Type:
        </label>
        <fieldset name="file_type" className="row-2 cell-7 md:cell-7">
          {Object.keys(fileTypeDetail).map(item => (
            <SelectElement
              key={item}
              value={item}
              select={e => setFileType(e.target.value)}
              selected={fileType === item}
            />
          ))}
        </fieldset>
        <label
          className="row-3 cell-3 text-right Form--Label Form--Label-required"
          htmlFor="description"
        >
          Describe document contents:
        </label>
        <textarea
          data-testid="description-input"
          className="Form--TextArea row-3 cell-7"
          type="text"
          name="description"
          placeholder={
            fileType && !fileDescription ? plachelderText[fileType].desc : ''
          }
          defaultValue={fileDescription}
          onChange={e => setFileDescription(e.target.value)}
        />
        <div className="row-4 cell-12 Form--Actions">
          {errors && <div className="text-red">{errors}</div>}
          <Button onClick={handleCancel}>Cancel</Button>
          <Button
            type="submit"
            color="primary"
            disabled={!fileName || !fileType || !fileDescription || onUploading}
          >
            {onUploading ? 'UPLOADING ...' : 'UPLOAD'}
          </Button>
        </div>
      </GridContainer>
    </form>
  );
};

NewFileForm.propTypes = {
  /** Function to perform on form submission */
  handleSubmit: PropTypes.func.isRequired,
  /** Function to perform when the cancel button is clicked */
  handleCancel: PropTypes.func.isRequired,
  /** Any errors that occured when submitting the form */
  errors: PropTypes.string,
};

export default NewFileForm;
