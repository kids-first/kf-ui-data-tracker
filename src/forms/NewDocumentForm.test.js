import React from 'react';
import {act} from 'react-dom/test-utils';
import {render} from 'react-testing-library';
import NewDocumentForm from './NewDocumentForm';

it('renders correctly', () => {
  const cancelMock = jest.fn();
  const submitMock = jest.fn();
  const tree = render(
    <NewDocumentForm handleCancel={cancelMock} handleSubmit={submitMock} />,
  );
  expect(tree).toMatchSnapshot();
});
