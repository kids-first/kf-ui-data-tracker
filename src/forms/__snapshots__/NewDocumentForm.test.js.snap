// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <form
        class="Form"
      >
        <section
          class="grid-container mx-auto grid-container--collapsed-rows"
        >
          <label
            class="row-1 cell-3 text-right Form--Label Form--Label-required"
            for="file_name"
          >
            Document Title:
          </label>
          <input
            class="Form--Input row-1 cell-7"
            data-testid="name-input"
            name="file_name"
            placeholder=""
            type="text"
            value=""
          />
          <label
            class="row-2 cell-3 text-right Form--Label Form--Label-required"
            for="file_type"
          >
            Document Type:
          </label>
          <fieldset
            class="row-2 cell-7 md:cell-7"
            name="file_type"
          >
            <label
              class="SelectElement"
            >
              <input
                class="border m-16"
                name="SHM"
                type="radio"
                value="SHM"
              />
              <div
                class="SelectElement--Icon"
              />
              <div>
                <p
                  class="m-0 font-bold font-sm"
                >
                  Shipping Manifest
                </p>
                <span
                  class="font-normal text-grey text-xs"
                >
                  File type description gose here...
                </span>
              </div>
            </label>
            <label
              class="SelectElement"
            >
              <input
                class="border m-16"
                name="CLN"
                type="radio"
                value="CLN"
              />
              <div
                class="SelectElement--Icon"
              />
              <div>
                <p
                  class="m-0 font-bold font-sm"
                >
                  Clinical/Phenotype Data
                </p>
                <span
                  class="font-normal text-grey text-xs"
                >
                  File type description gose here...
                </span>
              </div>
            </label>
            <label
              class="SelectElement"
            >
              <input
                class="border m-16"
                name="SEQ"
                type="radio"
                value="SEQ"
              />
              <div
                class="SelectElement--Icon"
              />
              <div>
                <p
                  class="m-0 font-bold font-sm"
                >
                  Sequencing Manifest
                </p>
                <span
                  class="font-normal text-grey text-xs"
                >
                  File type description gose here...
                </span>
              </div>
            </label>
            <label
              class="SelectElement"
            >
              <input
                class="border m-16"
                name="OTH"
                type="radio"
                value="OTH"
              />
              <div
                class="SelectElement--Icon"
              />
              <div>
                <p
                  class="m-0 font-bold font-sm"
                >
                  Other
                </p>
                <span
                  class="font-normal text-grey text-xs"
                >
                  File type description gose here...
                </span>
              </div>
            </label>
          </fieldset>
          <label
            class="row-3 cell-3 text-right Form--Label Form--Label-required"
            for="description"
          >
            Describe document contents:
          </label>
          <textarea
            class="Form--TextArea row-3 cell-7"
            data-testid="description-input"
            name="description"
            placeholder=""
            type="text"
          />
          <div
            class="row-4 cell-12 Form--Actions"
          >
            <button
              class="Button Button--default"
              type="button"
            >
              Cancel
            </button>
            <button
              class="Button Button--default Button--primary"
              disabled=""
              type="submit"
            >
              UPLOAD
            </button>
          </div>
        </section>
      </form>
    </div>
  </body>,
  "container": <div>
    <form
      class="Form"
    >
      <section
        class="grid-container mx-auto grid-container--collapsed-rows"
      >
        <label
          class="row-1 cell-3 text-right Form--Label Form--Label-required"
          for="file_name"
        >
          Document Title:
        </label>
        <input
          class="Form--Input row-1 cell-7"
          data-testid="name-input"
          name="file_name"
          placeholder=""
          type="text"
          value=""
        />
        <label
          class="row-2 cell-3 text-right Form--Label Form--Label-required"
          for="file_type"
        >
          Document Type:
        </label>
        <fieldset
          class="row-2 cell-7 md:cell-7"
          name="file_type"
        >
          <label
            class="SelectElement"
          >
            <input
              class="border m-16"
              name="SHM"
              type="radio"
              value="SHM"
            />
            <div
              class="SelectElement--Icon"
            />
            <div>
              <p
                class="m-0 font-bold font-sm"
              >
                Shipping Manifest
              </p>
              <span
                class="font-normal text-grey text-xs"
              >
                File type description gose here...
              </span>
            </div>
          </label>
          <label
            class="SelectElement"
          >
            <input
              class="border m-16"
              name="CLN"
              type="radio"
              value="CLN"
            />
            <div
              class="SelectElement--Icon"
            />
            <div>
              <p
                class="m-0 font-bold font-sm"
              >
                Clinical/Phenotype Data
              </p>
              <span
                class="font-normal text-grey text-xs"
              >
                File type description gose here...
              </span>
            </div>
          </label>
          <label
            class="SelectElement"
          >
            <input
              class="border m-16"
              name="SEQ"
              type="radio"
              value="SEQ"
            />
            <div
              class="SelectElement--Icon"
            />
            <div>
              <p
                class="m-0 font-bold font-sm"
              >
                Sequencing Manifest
              </p>
              <span
                class="font-normal text-grey text-xs"
              >
                File type description gose here...
              </span>
            </div>
          </label>
          <label
            class="SelectElement"
          >
            <input
              class="border m-16"
              name="OTH"
              type="radio"
              value="OTH"
            />
            <div
              class="SelectElement--Icon"
            />
            <div>
              <p
                class="m-0 font-bold font-sm"
              >
                Other
              </p>
              <span
                class="font-normal text-grey text-xs"
              >
                File type description gose here...
              </span>
            </div>
          </label>
        </fieldset>
        <label
          class="row-3 cell-3 text-right Form--Label Form--Label-required"
          for="description"
        >
          Describe document contents:
        </label>
        <textarea
          class="Form--TextArea row-3 cell-7"
          data-testid="description-input"
          name="description"
          placeholder=""
          type="text"
        />
        <div
          class="row-4 cell-12 Form--Actions"
        >
          <button
            class="Button Button--default"
            type="button"
          >
            Cancel
          </button>
          <button
            class="Button Button--default Button--primary"
            disabled=""
            type="submit"
          >
            UPLOAD
          </button>
        </div>
      </section>
    </form>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllBySelectText": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getAllByValue": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getBySelectText": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "getByValue": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllBySelectText": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryAllByValue": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryBySelectText": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "queryByValue": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
